File: server/src/main/java/yangFenTuoZi/runner/plus/server/Workarounds.java
package yangFenTuoZi.runner.plus.server;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Looper;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

@SuppressLint("PrivateApi,BlockedPrivateApi,SoonBlockedPrivateApi,DiscouragedPrivateApi")
public final class Workarounds {

    private static final Class<?> ACTIVITY_THREAD_CLASS;
    private static final Object ACTIVITY_THREAD;

    static {
        prepareMainLooper();

        try {
            // ActivityThread activityThread = new ActivityThread();
            ACTIVITY_THREAD_CLASS = Class.forName("android.app.ActivityThread");
            Constructor<?> activityThreadConstructor = ACTIVITY_THREAD_CLASS.getDeclaredConstructor();
            activityThreadConstructor.setAccessible(true);
            ACTIVITY_THREAD = activityThreadConstructor.newInstance();

            // ActivityThread.sCurrentActivityThread = activityThread;
            Field sCurrentActivityThreadField = ACTIVITY_THREAD_CLASS.getDeclaredField("sCurrentActivityThread");
            sCurrentActivityThreadField.setAccessible(true);
            sCurrentActivityThreadField.set(null, ACTIVITY_THREAD);
        } catch (Exception e) {
            throw new AssertionError(e);
        }
    }

    private Workarounds() {
        // not instantiable
    }

    private static void prepareMainLooper() {
        // Some devices internally create a Handler when creating an input Surface, causing an exception:
        //   "Can't create handler inside thread that has not called Looper.prepare()"
        // <https://github.com/Genymobile/scrcpy/issues/240>
        //
        // Use Looper.prepareMainLooper() instead of Looper.prepare() to avoid a NullPointerException:
        //   "Attempt to read from field 'android.os.MessageQueue android.os.Looper.mQueue'
        //    on a null object reference"
        // <https://github.com/Genymobile/scrcpy/issues/921>
        Looper.prepareMainLooper();
    }

    static Context getSystemContext() {
        try {
            Method getSystemContextMethod = ACTIVITY_THREAD_CLASS.getDeclaredMethod("getSystemContext");
            return (Context) getSystemContextMethod.invoke(ACTIVITY_THREAD);
        } catch (Throwable throwable) {
            // this is a workaround, so failing is not an error
            throwable.printStackTrace(System.err);
            return null;
        }
    }
}

File: app/src/main/java/yangfentuozi/runner/server/fakecontext/Workarounds.kt
package yangfentuozi.runner.server.fakecontext

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import yangfentuozi.runner.server.ServerMain
import java.lang.reflect.Constructor

@SuppressLint("PrivateApi,BlockedPrivateApi,SoonBlockedPrivateApi,DiscouragedPrivateApi")
object Workarounds {
    private var ACTIVITY_THREAD_CLASS: Class<*>? = null
    private var ACTIVITY_THREAD: Any? = null

    init {
        try {
            ACTIVITY_THREAD_CLASS = Class.forName("android.app.ActivityThread")
            val activityThreadConstructor: Constructor<*>? =
                ACTIVITY_THREAD_CLASS?.getDeclaredConstructor()
            activityThreadConstructor?.isAccessible = true
            ACTIVITY_THREAD = activityThreadConstructor?.newInstance()

            val sCurrentActivityThreadField =
                ACTIVITY_THREAD_CLASS?.getDeclaredField("sCurrentActivityThread")
            sCurrentActivityThreadField?.isAccessible = true
            sCurrentActivityThreadField?.set(null, ACTIVITY_THREAD)
        } catch (e: Exception) {
            throw AssertionError(e)
        }
    }

    val systemContext: Context?
        get() {
            try {
                val getSystemContextMethod =
                    ACTIVITY_THREAD_CLASS!!.getDeclaredMethod("getSystemContext")
                return getSystemContextMethod.invoke(ACTIVITY_THREAD) as Context?
            } catch (throwable: Throwable) {
                Log.e(
                    ServerMain.TAG,
                    "Workarounds: Failed to get system context: ${throwable.stackTraceToString()}",
                    throwable
                )
                return null
            }
        }
}